### DISCUSSION TITLE
Should a single-threaded server example program be added that actually reads the incoming body?

### DISCUSSION THREAD

**Discussion by elipp** [hyperium/hyper#3224] [General]
Hey,

as a hyper and Rust noobie, based on examples/single_threaded.rs, I spent a good while trying to modify the example to actually read the incoming request body, but got confused by the following issue:

```
error[E0277]: the trait bound `service::util::ServiceFn<[closure@src/main.rs:158:34: 158:44], Body>: Service<Request<hyper::body::Incoming>>` is not satisfied
   --> src/main.rs:164:18
    |
164 |                 .serve_connection(stream, service)
    |                  ^^^^^^^^^^^^^^^^ the trait `Service<Request<hyper::body::Incoming>>` is not implemented for `service::util::ServiceFn<[closure@src/main.rs:158:34: 158:44], Body>`
```

As it turns out, that's because my service_fn used a function that had the following signature:

`async fn handle_request(req: hyper::Request<Body>) -> http::Result<http::Response<String>>`,

where `Body` refers to the struct defined in the example. Instead, one had to use `req: hyper::Request<hyper::body::Incoming>` in order to satisfy that bound. As a new user, figuring out that I was supposed to use a different `Body` type than what we just defined was quite confusing.

For anyone wondering how the body is supposed to be read, it can be found in examples/web_api.rs:

`let whole_body = req.collect().await?.aggregate();`

Have a good day!

**Comment by seanmonstar** [comment]
The single threaded example is meant to show a very specific thing: that you _can_ use hyper with a local executor. It's somewhat of an advanced topic, where the other examples are meant to be show the other features more, like the echo or web_api examples.
